core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -169,35 +174,108 @@ protected PlanNodeStatsEstimate visitLogicalExpression(LogicalExpression node, V
-                    .sorted(Comparator.comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
-                    .findFirst();
+            for (Expression expression : groupedExpressions.keySet()) {
+                // calculate a combined estimate for each group so that we don't end up
+                // picking estimate from just one of the predicates in a BETWEEN clause
+                // or multiple range in-equalities on a symbol
+                List<Expression> expressions = groupedExpressions.get(expression);
+                PlanNodeStatsEstimate estimate = process(expressions.get(0));
+                if (!estimate.isOutputRowCountUnknown()) {
+                    for (int i = 1; i < expressions.size(); i++) {
+                        estimate = new FilterExpressionStatsCalculatingVisitor(estimate, session, types).process(expressions.get(i));
 
-            if (smallest.isEmpty()) {
-                return PlanNodeStatsEstimate.unknown();
+                        if (estimate.isOutputRowCountUnknown()) {
+                            break;
+                        }
+                    }
+                }
+                estimatesBuilder.add(estimate);
             }
+            return estimatesBuilder.build();
+        }
 
-            return smallest.get().mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+        private PlanNodeStatsEstimate intersectCorrelatedStats(PlanNodeStatsEstimate combinedEstimate, PlanNodeStatsEstimate termEstimate, double outputRowCount, TypeProvider types)
+        {
+            PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+            result.setOutputRowCount(outputRowCount);
+            // Update statistic range for symbols
+            concat(combinedEstimate.getSymbolsWithKnownStatistics().stream(), termEstimate.getSymbolsWithKnownStatistics().stream())
+                    .distinct()
+                    .forEach(symbol -> {
+                        SymbolStatsEstimate termSymbolStats = termEstimate.getSymbolStatistics(symbol);
+                        SymbolStatsEstimate combinedSymbolStats = combinedEstimate.getSymbolStatistics(symbol);
+
+                        StatisticRange termRange = StatisticRange.from(termSymbolStats);
+                        StatisticRange combinedRange = StatisticRange.from(combinedSymbolStats);
+
+                        StatisticRange intersect = termRange.intersect(combinedRange);
+                        result.addSymbolStatistics(symbol, SymbolStatsEstimate.buildFrom(combinedSymbolStats)
+                                .setStatisticsRange(intersect)
+                                .setNullsFraction(minExcludeNaN(combinedSymbolStats.getNullsFraction(), termSymbolStats.getNullsFraction()))
+                                .setAverageRowSize(averageExcludingNaNs(combinedSymbolStats.getAverageRowSize(), termSymbolStats.getAverageRowSize()))
```

Comment:
this should probably be some kind of weighted average, but I'm not sure what weighted means here